### PR TITLE
[Tests-Only] Files external php unit tests

### LINKED ISSUE #34735: Run files_external tests in Drone
Currently files_external tests are not run on Drone with the exception of the SMB suite and maybe another one.

We should run the tests there to make sure we can catch errors especially during library updates.
For example, when upgrading phpseclib we need to make sure that the SFTP suite still works. (@phil-davis remembers)

Here is a PR that attempts to enable Webdav with OC and lists the failures: https://github.com/owncloud/core/pull/30733

List of storages to cover:
- [x] smb_samba
- [x] webdav_apache
- [x] webdav_owncloud #36857 
- [x] swift
- [x] SFTP #36292 
- [x] Amazon S3 - testing with scality
- Dropbox => currently not possible, see https://github.com/owncloud/files_external_dropbox/pull/17
- GDrive => moved to separate repo on master

### PR DESCRIPTION
**Pull request by kiranparajuli589**
## Description
Runs files external "owncloud" tests on drone

## Related Issue
- Fixes https://github.com/owncloud/core/issues/34735

## Motivation and Context
Better test coverage

## How Has This Been Tested?
- :robot: 

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests only (no source changes)

## Checklist:
- [x] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

**Comment by mmattel** [comment]
@phil-davis ping, sonarcloud seems to be stuck...

File: .drone.star [hunk 1]
```diff
@@ -2393,6 +2114,95 @@ def owncloudLog(server, folder):
...
+		'pull': 'always',
+		'environment': {
+			'MARIADB_ROOT_PASSWORD': 'owncloud',
+			'MARIADB_USERNAME': 'owncloud',
+			'MARIADB_PASSWORD': 'owncloud',
+			'MARIADB_DATABASE': 'owncloud',
+			'MARIADB_MAX_ALLOWED_PACKET': '128M',
+			'MARIADB_INNODB_LOG_FILE_SIZE': '64M'
+		}
+	}]
```
Comment (dpakach) [inline .drone.star:2145]: there is already function for `databaseService` no need to redefine this here.

File: .drone.star [hunk 2]
```diff
@@ -2393,6 +2114,95 @@ def owncloudLog(server, folder):
...
+
+def fileExternalPhpUnitTests(ctx):
+	return {
+		'kind': 'pipeline',
+		'type': 'docker',
+		'name': 'FilesExternalUnitTests',
+		'steps':
+			runPhpTest()
+		,
```
Comment (dpakach) [inline .drone.star:2190]:
```suggestion
		'steps': runPhpTest(),
```

File: .drone.star [hunk 3]
```diff
@@ -2393,6 +2114,95 @@ def owncloudLog(server, folder):
...
+		'services':
+			redisServer() +
+			mariaDBServer() +
+			owncloudServer()
+		,
```
Comment (dpakach) [inline .drone.star:2195]:
```suggestion
		'services': redisServer() +
			mariaDBServer() +
			owncloudServer() ,
```

File: .drone.star [hunk 4]
```diff
@@ -2393,6 +2114,95 @@ def owncloudLog(server, folder):
...
+	return [{
+		'name': 'oc-server',
+		'image': 'owncloud/server',
+		'pull': 'always',
+		'environment': {
+			'OWNCLOUD_VERSION': '10.6',
+			'OWNCLOUD_DOMAIN': 'oc-server',
+			'ADMIN_USERNAME': 'admin',
+			'ADMIN_PASSWORD': 'admin',
+			'HTTP_PORT': '8080'
```
Comment (dpakach) [inline .drone.star:2127]: You can maybe run this on port 80 and it can be accessed at `http://oc-server`

File: .drone.star [hunk 5]
```diff
@@ -19,289 +19,10 @@ config = {
...
-
-	dependsOn(before, stages)
-
-	afterCoverageTests = afterCoveragePipelines(ctx)
-	dependsOn(coverageTests, afterCoverageTests)
-
-	after = afterPipelines(ctx)
-	dependsOn(afterCoverageTests + stages, after)
-
-	return initial + before + coverageTests + afterCoverageTests + stages + after
```
Comment (dpakach) [inline .drone.star:304]: restore all the deleted code

**Review by phil-davis — COMMENTED** [review]
https://drone.owncloud.com/owncloud/core/28789
`unknown binary op: list + dict `

There is a starlark problem. Sadly it gives no traceback with line number!

File: .drone.star [hunk 6]
```diff
@@ -2393,6 +2394,85 @@ def owncloudLog(server, folder):
...
+		'trigger': {
+			'ref': [
+				'refs/pull/**',
+				'refs/tags/**'
+			]
+		}
+	}]
+
+def fileExternalPhpUnitTests():
+	return {
```
Comment (phil-davis) [inline .drone.star:2458]:
```suggestion
	return [{
```

File: .drone.star [hunk 7]
```diff
@@ -2393,6 +2394,85 @@ def owncloudLog(server, folder):
...
+			databaseService('mariadb:10.5') +
+			owncloudDockerServer(),
+		'depends_on': [],
+		'trigger': {
+			'ref': [
+				'refs/pull/**',
+				'refs/tags/**'
+			]
+		}
+	}
```
Comment (phil-davis) [inline .drone.star:2473]:
```suggestion
	}]
```

**Review by phil-davis — COMMENTED** [review]
I suspect that this needs `[]` around it to make the "dict" be one element of a list of pipelines.

File: .drone.star [hunk 8]
```diff
@@ -2393,6 +2394,85 @@ def owncloudLog(server, folder):
...
+			'OWNCLOUD_DB_NAME': getDbDatabase('mariadb'),
+      'OWNCLOUD_DB_USERNAME': getDbUsername('mariadb'),
+      'OWNCLOUD_DB_PASSWORD': getDbPassword('mariadb'),
+      'OWNCLOUD_DB_HOST': getDbName('mariadb'),
+			'OWNCLOUD_ADMIN_USERNAME': 'admin',
+			'OWNCLOUD_ADMIN_PASSWORD': 'admin',
+			'HTTP_PORT': '8080',
+			'OWNCLOUD_MYSQL_UTF8MB4': 'true',
+      'OWNCLOUD_REDIS_ENABLED': 'true',
+      'OWNCLOUD_REDIS_HOST': 'redis-docker-server'
```
Comment (phil-davis) [inline .drone.star:2415]: Indent is not consistent.

File: .drone.star [hunk 9]
```diff
@@ -313,11 +313,12 @@ def coveragePipelines(ctx):
 	# All pipelines that might have coverage or other test analysis reported
 	jsPipelines = javascript(ctx)
 	phpUnitPipelines = phpTests(ctx, 'phpunit')
+	phpUnitFilesExternalPipeline = fileExternalPhpUnitTests()
```
Comment (phil-davis) [inline .drone.star:316]:
```suggestion
	phpUnitFilesExternalPipeline = filesExternalPhpUnitTests()
```

File: .drone.star [hunk 10]
```diff
@@ -2393,6 +2394,85 @@ def owncloudLog(server, folder):
...
+		'depends_on': [],
+		'trigger': {
+			'ref': [
+				'refs/pull/**',
+				'refs/tags/**'
+			]
+		}
+	}]
+
+def fileExternalPhpUnitTests():
```
Comment (phil-davis) [inline .drone.star:2457]:
```suggestion
def filesExternalPhpUnitTests():
```

File: .drone.star [hunk 11]
```diff
@@ -2393,6 +2394,85 @@ def owncloudLog(server, folder):
...
+	return [{
+		'name': 'redis-docker-server',
+		'image': 'webhippie/redis:latest',
+		'pull': 'always',
+		'environment': {
+			'REDIS_DATABASES': 1
+		}
+	}]
+
+def runPhpTest():
```
Comment (phil-davis) [inline .drone.star:2431]:
```suggestion
def runFilesExternalPhpUnitTests():
```

File: .drone.star [hunk 12]
```diff
@@ -2393,6 +2394,85 @@ def owncloudLog(server, folder):
...
+	return [{
+		'name': 'files-external-unit-tests',
+		'image': 'owncloudci/php:7.4',
+		'pull': 'always',
+		'environment': {
+			'TEST_PHP_SUITE': 'apps/files_external/tests/Storage/OwncloudTest.php'
```
Comment (phil-davis) [inline .drone.star:2437]: https://drone.owncloud.com/owncloud/core/28792/34/5 ran the whole 10,000-test PHP unit tests.
Maybe this has to be specified on the `make` command line?

File: .drone.star [hunk 13]
```diff
@@ -2393,6 +2394,85 @@ def owncloudLog(server, folder):
...
+		},
+		'commands': [
+			'wait-for-it -t 1200 oc-docker-server:8080 -- echo "OC server is up."',
+			'sed -i "48 s/false/true/" apps/files_external/tests/config.php',
+			'sed -i "s/localhost\\\/owncloud/oc-server/g" apps/files_external/tests/config.php',
+			'cd /drone/src',
+			'make',
+			'curl -u admin:admin http://oc-docker-server:8080/ocs/v1.php/cloud/users -d userid="test" -d password="test"',
+			'make test-php-unit'
```
Comment (phil-davis) [inline .drone.star:2446]:
```suggestion
			'make test-php-unit TEST_PHP_SUITE=apps/files_external/tests/Storage/OwncloudTest.php'
```

Maybe you have to do this?

**Comment by kiranparajuli589** [comment]
@phil-davis @dpakach addressed your comments. please review again.

**Review by dpakach — COMMENTED** [review]
LGTM

**Comment by kiranparajuli589** [comment]
at the date of the comment...blocked because docker `owncloud/server` is not working properly.

**Comment by phil-davis** [comment]
https://drone.owncloud.com/owncloud/core/30001/34/11
```
PHPUnit 8.5.15 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.2.34-21+ubuntu18.04.1+deb.sury.org+1
Configuration: /drone/src/tests/phpunit-autotest.xml
Error:         No code coverage driver is available

.............................................F................... 65 / 85 ( 76%)
....................                                              85 / 85 (100%)

Time: 2.21 minutes, Memory: 28.00 MB

There was 1 failure:

1) OCA\Files_External\Tests\Storage\OwncloudTest::testCheckUpdate
No update
Failed asserting that true is false.

/drone/src/tests/lib/Files/Storage/Storage.php:347
```

This looks like a real fail to me. 
```
	public function testCheckUpdate() {
		if ($this->instance instanceof \OC\Files\Storage\Wrapper\Wrapper) {
			$this->markTestSkipped('Cannot test update check on wrappers');
		}
		$textFile = \OC::$SERVERROOT . '/tests/data/lorem.txt';
		$watcher = $this->instance->getWatcher();
		$watcher->setPolicy(Watcher::CHECK_ALWAYS);
		$this->instance->file_put_contents('/lorem.txt', \file_get_contents($textFile));
		$this->assertTrue($watcher->checkUpdate('/lorem.txt'), 'Update detected');
		$this->assertFalse($watcher->checkUpdate('/lorem.txt'), 'No update');
	}
```
`checkUpdate` is supposed to check the local cache and if it is new/changed then update the storage. A second call to `checkUpdate` should find that no update is needed.

**Comment by phil-davis** [comment]
https://drone.owncloud.com/owncloud/core/30013/31/10
It needs `tests/drone/configs/config.files_external.owncloud.php` to be created. That seems to be called `config.files_external_owncloud.php` at the moment.

**Comment by kiranparajuli589** [comment]
> https://drone.owncloud.com/owncloud/core/30001/34/11
> 
> ```
> PHPUnit 8.5.15 by Sebastian Bergmann and contributors.
> 
> Runtime:       PHP 7.2.34-21+ubuntu18.04.1+deb.sury.org+1
> Configuration: /drone/src/tests/phpunit-autotest.xml
> Error:         No code coverage driver is available
> 
> .............................................F................... 65 / 85 ( 76%)
> ....................                                              85 / 85 (100%)
> 
> Time: 2.21 minutes, Memory: 28.00 MB
> 
> There was 1 failure:
> 
> 1) OCA\Files_External\Tests\Storage\OwncloudTest::testCheckUpdate
> No update
> Failed asserting that true is false.
> 
> /drone/src/tests/lib/Files/Storage/Storage.php:347
> ```
> 
> This looks like a real fail to me.
> 
> ```
> 	public function testCheckUpdate() {
> 		if ($this->instance instanceof \OC\Files\Storage\Wrapper\Wrapper) {
> 			$this->markTestSkipped('Cannot test update check on wrappers');
> 		}
> 		$textFile = \OC::$SERVERROOT . '/tests/data/lorem.txt';
> 		$watcher = $this->instance->getWatcher();
> 		$watcher->setPolicy(Watcher::CHECK_ALWAYS);
> 		$this->instance->file_put_contents('/lorem.txt', \file_get_contents($textFile));
> 		$this->assertTrue($watcher->checkUpdate('/lorem.txt'), 'Update detected');
> 		$this->assertFalse($watcher->checkUpdate('/lorem.txt'), 'No update');
> 	}
> ```
> 
> `checkUpdate` is supposed to check the local cache and if it is new/changed then update the storage. A second call to `checkUpdate` should find that no update is needed.

@phil-davis  we can skip this with a new issue

File: apps/files_external/tests/Storage/OwncloudTest.php [hunk 14]
```diff
@@ -41,7 +41,7 @@ protected function setUp(): void {
 		parent::setUp();
 
 		$id = $this->getUniqueID();
-		$this->config = include('files_external/tests/config.php');
+		$this->config = include('files_external/tests/config.owncloud.php');
```
Comment (phil-davis) [inline apps/files_external/tests/Storage/OwncloudTest.php:44]: IMO `files_external/tests/config.owncloud.php` does not exist. The config for ownCloud is already in `files_external/tests/config.php'

Comment (kiranparajuli589) [reply apps/files_external/tests/Storage/OwncloudTest.php:44]: resolved

**Comment by phil-davis** [comment]
> we can skip this with a new issue

I created issue #38371 and pushed some code that should skip the test.

**Comment by phil-davis** [comment]
https://drone.owncloud.com/owncloud/core/30022/31/10
```
PHPUnit 8.5.15 by Sebastian Bergmann and contributors.

Runtime:       PHPDBG 7.2.34-21+ubuntu18.04.1+deb.sury.org+1
Configuration: /drone/src/tests/phpunit-autotest-external.xml

.............................................S................... 65 / 85 ( 76%)
....................                                              85 / 85 (100%)

Time: 2.01 minutes, Memory: 52.52 MB

There was 1 skipped test:

1) OCA\Files_External\Tests\Storage\OwncloudTest::testCheckUpdate
Skip testCheckUpdate on OwncloudTest see issue 38371

/drone/src/tests/lib/Files/Storage/Storage.php:343

OK, but incomplete, skipped, or risky tests!
Tests: 85, Assertions: 364, Skipped: 1.
```

Looks good - the test is skipped.

**Review by phil-davis — APPROVED** [review]

**Pull request merged**